File: Algorithms/src/3346. Maximum Frequency of an Element After Performing Operations I/Code/solution.go
func maxFrequency(nums []int, k int, numOperations int) int {
    if len(nums) == 0 {
        return 0
    }
    mx := nums[0]
    for _, v := range nums {
        if v > mx {
            mx = v
        }
    }
    size := mx + k + 2
    count := make([]int, size)
    for _, v := range nums {
        count[v]++
    }
    for i := 1; i < size; i++ {
        count[i] += count[i-1]
    }

    ans := 0
    for t := 0; t < size; t++ {
        L := t - k
        if L < 0 {
            L = 0
        }
        R := t + k
        if R > size-1 {
            R = size-1
        }
        total := count[R]
        if L > 0 {
            total -= count[L-1]
        }
        freq_t := count[t]
        if t > 0 {
            freq_t -= count[t-1]
        }
        canConvert := total - freq_t
        take := numOperations
        if take > canConvert {
            take = canConvert
        }
        val := freq_t + take
        if val > ans {
            ans = val
        }
    }
    return ans
}

File: Algorithms/src/3346. Maximum Frequency of an Element After Performing Operations I/Code/solution.py
class Solution:
    def maxFrequency(self, nums: List[int], k: int, numOperations: int) -> int:
        if not nums:
            return 0
        mx = max(nums)
        size = mx + k + 2
        count = [0] * size

        # frequency array
        for v in nums:
            count[v] += 1

        # prefix sums
        for i in range(1, size):
            count[i] += count[i-1]

        ans = 0
        for t in range(size):
            L = max(0, t - k)
            R = min(size - 1, t + k)
            total = count[R] - (count[L-1] if L > 0 else 0)   # elements that can become t
            freq_t = count[t] - (count[t-1] if t > 0 else 0) # how many already t
            canConvert = total - freq_t
            take = min(numOperations, canConvert)
            ans = max(ans, freq_t + take)
        return ans
